src/fstext/kaldi-fst-io.cc

```diff
@@ -42,6 +42,67 @@ VectorFst<StdArc> *ReadFstKaldi(std::string rxfilename) {
   return fst;
 }
 
+Fst<StdArc> *ReadFstKaldiGeneric(std::string rxfilename, bool throw_on_err) {
+  if (rxfilename == "") rxfilename = "-"; // interpret "" as stdin,
+  // for compatibility with OpenFst conventions.
+  kaldi::Input ki(rxfilename);
+  fst::FstHeader hdr;
+  // Read FstHeader which contains the type of FST
+  if (!hdr.Read(ki.Stream(), rxfilename)) {
+    if(throw_on_err) {
+      KALDI_ERR << "Reading FST: error reading FST header from "
+                << kaldi::PrintableRxfilename(rxfilename);
+    } else {
+      std::cout << "Warning: We fail to read FST header from "
+                << kaldi::PrintableRxfilename(rxfilename) 
+                << ". A NULL pointer is returned." << std::endl;
+      return NULL;
+    }
+  }
+  // Check the type of Arc
+  if (hdr.ArcType() != fst::StdArc::Type()) {
+    if(throw_on_err) {
+      KALDI_ERR << "FST with arc type " << hdr.ArcType() << " is not supported.";
+    } else {
+      std::cout << "Warning: Fst with arc type" << hdr.ArcType()
+                << " is not supported. A NULL pointer is returned." << std::endl;
+      return NULL;
+    }
+  }
+  // Read the FST
+  FstReadOptions ropts("<unspecified>", &hdr);
+  Fst<StdArc> *fst = NULL;
+  if (hdr.FstType() == "const") {
+    fst = ConstFst<StdArc>::Read(ki.Stream(), ropts);
+  } else if (hdr.FstType() == "vector") {
+    fst = VectorFst<StdArc>::Read(ki.Stream(), ropts);
+  }
+  if (!fst) {
+    if(throw_on_err) {
+     KALDI_ERR << "Could not read fst from "
+               << kaldi::PrintableRxfilename(rxfilename);
+    } else {
+      std::cout << "Warning: Could not read fst from "
+                << kaldi::PrintableRxfilename(rxfilename)
+                << ". A NULL pointer is returned." << std::endl;
+      return NULL;
+    }
+  }
+  return fst;
+}
+
+VectorFst<StdArc> *CastToVectorFst(Fst<StdArc> *fst) {
+  std::string real_type = fst->Type();
+  if (real_type == "vector") {
+    return dynamic_cast<VectorFst<StdArc> *>(fst);
+  } else {
```

Comment:
Probably a good idea to have a KALDI_ASSERT here like
 KALDI_ASSERT(real_type == "const" && real_type == "some other type that it supports");

so that if it's a type that isn't supported the assertion would fail